ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsRange.scala

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.daml.platform.store.dao.events
+
+import anorm.SqlParser.get
+import anorm.SqlStringInterpolation
+import com.daml.ledger.participant.state.v1.Offset
+
+final case class EventsRange[A](startExclusive: A, endInclusive: A)
+
+object EventsRange {
+  private val EmptyRowIdRange = EventsRange(0L, 0L)
+
+  private val oneRow = Some(1)
+
+  /**
+    * Converts Offset range to Row ID range.
+    *
+    * @param range offset range
+    * @param connection SQL connection
+    * @return Row ID range
+    */
+  def readRowIdRange(range: EventsRange[Offset])(
+      connection: java.sql.Connection): EventsRange[Long] =
+    EventsRange(
+      startExclusive = readLowerBound(range.startExclusive)(connection),
+      endInclusive = readUpperBound(range.endInclusive)(connection)
+    )
+
+  /**
+    * Converts ledger end offset into a Row ID range.
+    *
+    * @param endInclusive ledger end offset
+    * @param connection SQL connection
+    * @return Row ID range
+    */
+  def readRowIdRange(endInclusive: Offset)(connection: java.sql.Connection): EventsRange[Long] =
+    EmptyRowIdRange.copy(endInclusive = readUpperBound(endInclusive)(connection))
+
+  private def readLowerBound(startExclusive: Offset)(connection: java.sql.Connection): Long =
+    if (startExclusive == Offset.beforeBegin) {
+      EmptyRowIdRange.startExclusive
+    } else {
+      import com.daml.platform.store.Conversions.OffsetToStatement
+      // This query could be: "select min(row_id) - 1 from participant_events where event_offset > ${range.startExclusive}"
+      // however there are cases when postgres decides not to use the index. We are forcing the index usage specifying `order by event_offset`
```

Comment:
I'm starting to wonder: was the query slow just because PostgreSQL was not picking the right index?

Reply:
don't know... however `order by event_sequential_id` is faster than `(event_offset, transaction_id, node_index)` that is for sure. Also `event_sequential_id` allows us to simplify the SQL where clause, we don't have to take into account `node_index`, because `event_sequential_id` is an unique identifier.